hazelcast/src/main/java/com/hazelcast/client/impl/protocol/ClientMessage.java

```diff
@@ -137,47 +139,73 @@ private ClientMessage() {
 
     }
 
-    private ClientMessage(LinkedList<Frame> frames) {
-        super(frames);
+    private ClientMessage(Frame startFrame) {
+        this.startFrame = startFrame;
+        endFrame = startFrame;
+        while (endFrame.next != null) {
+            endFrame = endFrame.next;
+        }
     }
 
     public static ClientMessage createForEncode() {
         return new ClientMessage();
     }
 
-    public static ClientMessage createForDecode(LinkedList<Frame> frames) {
-        return new ClientMessage(frames);
+    public static ClientMessage createForDecode(Frame startFrame) {
+        return new ClientMessage(startFrame);
+    }
+
+    public Frame getStartFrame() {
+        return startFrame;
+    }
+
+    public ClientMessage add(Frame frame) {
+        frame.next = null;
+        if (startFrame == null) {
+            startFrame = frame;
+            endFrame = frame;
+            return this;
+        }
+
+        endFrame.next = frame;
+        frame.previous = endFrame;
+        endFrame = frame;
+        return this;
+    }
+
+    public FrameIterator frameIterator() {
+        return new FrameIterator(startFrame);
```

Comment:
Is it possible to remove FrameIterator also? ClientMessage could be the iterator itself. The drawback is you can not have two different iterators but I think we don't need it anyway.

Reply:
OK, changed.

Reply:
It came out that this was not a good idea. Some tests failed and when examined, I found out that the ClientMessage decode may happen from different executor threads and hence it is not thread safe if we keep the nextFrame variable in the ClientMessage which is being modified and read through different threads. Failing test: testGetStat_whenGetAsync  This line https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/ClientDelegatingFuture.java#L91  the `DelegatingExecutionCallback` may decode the message using different executors.

Hence, I am reverting back the changes.